### ISSUE TITLE
max_unpool2d and max_unpool3d cpp signature should be similar

### ISSUE THREAD

**Issue by ailzhang** [pytorch/pytorch#39757]
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L6021-L6047
Currently we have different signatures for `max_unpool2d` and `max_unpool3d`.
```
- func: max_unpool2d(Tensor self, Tensor indices, int[2] output_size) -> Tensor
- func: max_unpool3d(Tensor self, Tensor indices, int[3] output_size, int[3] stride, int[3] padding) -> Tensor
```
From python API, it looks like we should move the `calculate output_size` logic to cpp and make `max_unpool2d` signature similar to `max_unpool3d`. currently stride/padding are passed to cpp but not used. 
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L706

**Comment by chengjinfang1216** [comment]
@alizhang, I'd like to work on this issue, is that ok?

**Comment by ailzhang** [comment]
Hey @chengjinfang1216! Of course, let us know if you need any help!

**Comment by chengjinfang1216** [comment]
hello, @ailzhang, I found python method  max_unpool1d, it calls torch._C._nn.max_unpool2d to implement the logic, but some check rules(for example: stride.size should equal to 2) only apply to 2d tensor.  How about adding new methods for 1d tensor, it may check the input values like below
`  check1d("max_pool1d", "stride", stride);`
Then we can call max_unpool2d related method.
could you please offer me some advice? thanks a lot.

**Comment by ailzhang** [comment]
Hi @chengjinfang1216, yea that makes sense to me!

**Comment by chengjinfang1216** [comment]
>From python API, it looks like we should move the `calculate output_size` logic to cpp 
Hi, @ailzhang, about the comment you pointed out,  I checked the CPP API and found that it also contains a similar logic.
Did you mean that we should move this method to MaxUnpooling.cpp and MaxUnpooling.cu files so that Both Python and CPP API  will call the same method?

**Comment by ailzhang** [comment]
@chengjinfang1216 Nope the logic is only needed once so we'd prefer to keep the CPP one :D It should be fine to remove the logic on python side. Thanks!

**Comment by chengjinfang1216** [comment]
Hi @alizhang, about this I want to confirm with you again.
In current source implementation，  Python API  is as follows：https://github.com/pytorch/pytorch/blob/60f2fa6a8422f138c2e7eaa302f04bf32a48c02b/torch/nn/functional.py#L706
while CPP API is as follows：
https://github.com/pytorch/pytorch/blob/60f2fa6a8422f138c2e7eaa302f04bf32a48c02b/torch/csrc/api/include/torch/nn/functional/pooling.h#L681
If we only remove the python side logic and keep the cpp one,  when user use python API, it will not call _unpool_output_size in CPP side.
So I think we should move the call logic to the following methods(max_unpooling2d_forward_cpu, max_unpooling2d_forward_cuda) declared in native_functions.yaml, if so, both python and CPP can call the same logic.
Could you please confirm for me or is there any better implementation?
https://github.com/pytorch/pytorch/blob/60f2fa6a8422f138c2e7eaa302f04bf32a48c02b/aten/src/ATen/native/native_functions.yaml#L5958-L5963

**Comment by alizhang** [comment]
> Hi @alizhang, about this I want to confirm with you again.
> In current source implementation， Python API is as follows：
> 
> https://github.com/pytorch/pytorch/blob/60f2fa6a8422f138c2e7eaa302f04bf32a48c02b/torch/nn/functional.py#L706
> 
> 
> while CPP API is as follows：
> https://github.com/pytorch/pytorch/blob/60f2fa6a8422f138c2e7eaa302f04bf32a48c02b/torch/csrc/api/include/torch/nn/functional/pooling.h#L681
> 
> 
> If we only remove the python side logic and keep the cpp one, when user use python API, it will not call _unpool_output_size in CPP side.
> So I think we should move the call logic to the following methods(max_unpooling2d_forward_cpu, max_unpooling2d_forward_cuda) declared in native_functions.yaml, if so, both python and CPP can call the same logic.
> Could you please confirm for me or is there any better implementation?
> https://github.com/pytorch/pytorch/blob/60f2fa6a8422f138c2e7eaa302f04bf32a48c02b/aten/src/ATen/native/native_functions.yaml#L5958-L5963

Hi, it seems that you want to reply to @ailzhang, not me :D

**Comment by ailzhang** [comment]
@chengjinfang1216 Please correct me if I'm wrong, 
> If we only remove the python side logic and keep the cpp one, when user use python API, it will not call _unpool_output_size in CPP side.
I thought `_unpool_output_size` are doing the same thing so it's safe to remove the python one?

**Comment by chengjinfang1216** [comment]
> @chengjinfang1216 Please correct me if I'm wrong,
> 
> > If we only remove the python side logic and keep the cpp one, when user use python API, it will not call _unpool_output_size in CPP side.
> > I thought `_unpool_output_size` are doing the same thing so it's safe to remove the python one?

@alizhang , Yep, there are two separate _unpool_output_size functions in both Python and CPP codes and doing the same thing.

**Comment by jinchen62** [comment]
Hi, can someone help on this issue? The PR got closed https://github.com/pytorch/pytorch/pull/40155.

**Comment by vivekkhandelwal1** [comment]
Hi @Chillee @ngimel, can you please help us on this issue? Or, maybe assign this to someone who can help?
